### ISSUE TITLE
AWS ECR Authorization not recognized after Docker Login

### ISSUE THREAD

**Issue by Jawery** [kubewarden/kwctl#208]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When trying to interact with aws ECR to push or pull an artifact I'm getting a 401 unauthorized even after a successful `docker login` and display of ability to push/pull normal docker images. Verified a few things over on the slack channel so I'll repeat them here for simplicity:

- Logging in to docker from the cli piping the result from `aws ecr get-login-password` to the login command.
- Verified I was using a token that wasn't expired
- Tried passing in the .docker/config.json file itself with no change
- Same functionality on v0.2.5 and v0.3.0

### Expected Behavior

Authentication through docker cli to aws ecr enables kwctl interaction with aws ecr as well. Alternatively some other auth solution for aws would work as well.

### Steps To Reproduce

1. `Docker Login` to aws ecr
2. Verify that `Docker push/pull [some ecr repo]` is successful after authentication
4. `kwctl push [some ecr repo]`
5. Here I'm encountering a 401 response from aws

### Environment

```markdown
- OS: Mac
- Architecture:x86-64
```

**Comment by viccuad** [comment]
As a workaround, it should be possible to use [`wasm-to-oci`](https://github.com/engineerd/wasm-to-oci), which lists aws ECR. An  example pushing to GH packages:
```
$ git clone git@github.com:kubewarden/pod-privileged-policy.git
$ cd pod-privileged-policy
$ make annotated-policy.wasm
$ wasm-to-oci push annotated-policy.wasm ghcr.io/viccuad/policies/pod-privileged:wasm-to-oci
INFO[0005] Pushed: ghcr.io/viccuad/policies/pod-privileged:wasm-to-oci
INFO[0005] Size: 21892
INFO[0005] Digest: sha256:4ada0a5ceda88ef69f26ab11213e99f5595465442ed3b5665a868ac3d919ef0e
```

**Comment by Jawery** [comment]
As mentioned in the slack channel unfortunately it seems `wasm-to-oci` isn't working with AWS ECR currently, when the policyServer tries to pull an artifact uploaded using this tool from AWS ECR the logs display an error `policy could not be downloaded due to error: error decoding response body: expected value at line 1 column 1`

I have however discovered that the alternative of embedding my .wasm in the policyServer image and referencing the policy with the `file://[PATH]` format works just fine so we now have a workaround that isn't too cumbersome. We would of course prefer to function in AWS but no longer are we blocked by this issue.

**Comment by flavio** [comment]
@Jawery a better solution would be to fetch the policy from a http/https server. I'll look into the issue

**Comment by flavio** [comment]
I can reproduce the issue. I'm going to look into that.

**Comment by flavio** [comment]
This issue is fixed starting from kwctl release v0.3.3. The binaries are being built by the CI as I speak. They are going to be available soon for download here: https://github.com/kubewarden/kwctl/releases

**Comment by Jawery** [comment]
@flavio Thanks so much for your work on this, aws not following oci specs is a pretty interesting situation. I downloaded kwctl 3.3 and am still getting the same 401 unauthorized however (which upon inspection sounds like a different issue than you've fixed relating to oci specs). Something additional that I hadn't noticed before is that the error says ` Could not push policy: Server Error: https://{aws registry URI}}/blobs/uploads/ Code: 401 Unauthorized` , specifically the pathing to /blobs/uploads/ seems to be potentially incorrect, as even in the browser this url returns a 401 after I've federated into my account. Considering that these same credentials function for pushing both standard docker images and arbitrary binaries (wasm-to-oci/oras used to confirm) I'm confident it isn't an aws permissions thing and maybe it's just the wrong endpoint. Any thoughts on this idea?

**Comment by flavio** [comment]
Ouch, you're right. The fix isn't working as expected. I already prepared a new fix and successfully tested that inside of an experimental build of kwct.

I'll tag a new release of kwctl once https://github.com/krustlet/oci-distribution/pull/40 gets merged and released (hopefully later today)

**Comment by Jawery** [comment]
@flavio Hey sorry, I was out for a few days and just got back in. I'm still getting the 401 unauthorized error with kwctl v0.3.5 when following the steps in the original issue post. It still seems to me that the issue is perhaps unrelated to the oci spec violation/workaround you've mentioned. Like I said in my last message the error prints out `https://{aws registry URI}}/blobs/uploads/` after entering just `registry://{aws registry URI}}` and navigating to the `/blobs/uplaods` url in a browser (that I'm federated in) gives a 401 as well giving me the impression that it's just the wrong address. Any thoughts?

**Comment by raulcabello** [comment]
I can't reproduce this issue with [kwctl 1.0.1](https://github.com/kubewarden/kwctl/releases/tag/v1.0.1). Everything is working as expected. This is what I did:

1. Logging into docker using `aws ecr get-login-password --region eu-central-1 | docker login --username AWS --password-stdin xxxxxxx.ecr.eu-central-1.amazonaws.com`
2. Then `kwctl push annotated-policy.wasm registry://xxxx.ecr.eu-central-1.amazonaws.com/xxxx` works as expected
3. And `kwctl pull` also works as expected

@Jawery Can you please provide more info so I can reproduce this? Am I doing something different than you?

It looks like aws is using the `https://{aws registry URI}}/blobs/uploads/ `url internally. This is the error I see if I remove permission to my aws user for the ecr repository:

```
Error: could not push policy: could not push policy: Server error: url https://xxxxxx.ecr.eu-central-1.amazonaws.com/v2/blobs/uploads/xxxxxxxxxxxxxxx, code: 403, message: {"errors":[{"code":"DENIED","message":"User: arn:aws:iam::xxxxxxx is not authorized to perform: ecr:UploadLayerPart on resource: arn:aws:ecr:eu-central-1:xxxxxxxxx because no identity-based policy allows the ecr:UploadLayerPart action"}]}
```

**Comment by Jawery** [comment]
This is very strange to me, I'm following the exact steps you are, the only difference being a different aws region (us-east-1) and the format of the registry url I'm using has `dkr` added to it  `registry://xxxx.dkr.ecr.us-east-1.amazonaws.com/xxxx`. (Thought this was notable due to the similarity to 'docker'). I'm wondering if this may have something to do with it, but removing it I get a dns error. 

Also the error I get isn't the same as yours, since I get a 401 rather than the 403 you're getting, so it's at least not the same permission problem you recreated. Another note too, between steps 2 and 3 in your list I'm able to do docker pull/pushes without issue, indicating that aws permissions are set and I'm successfully signing in. 

My current thoughts as to what might be going on are:

1. That .`dkr` component might be playing a role, unless yours is there as well just grouped with the account number x's.
2. Some difference with the docker config file maybe? Mine is configured using the ootb config, so it's comprised as a list of auth endpoints mapping to empty sets, and a separate field `"credStore": "Desktop"` that presumably is used to retrieve the non-plaintext creds. I'm not particularly familiar with this element of docker, but if we have no differences elsewhere this might be a lead.

**Comment by raulcabello** [comment]
@Jawery I finally managed to reproduce the issue! It looks like this is happening just on mac (I was testing on linux before)
This is what I see in my mac:
```
Error: could not push policy: could not push policy: Server error: url https://xxxxxxx.dkr.ecr.eu-central-1.amazonaws.com/v2/test/blobs/uploads/, code: 401, message: Not Authorized
```

I'll look into that, and I'll let you know once I have more details about this.

**Comment by raulcabello** [comment]
This is happening because mac is using the docker credential store (if you are using Docker Desktop). Currently `kwctl` assumes the credentials are stored in plan text in the `.docker/config.json`
We have to read the credential from the credentials store if the field `credsStore` is in the `.docker/config.json`.
I did a proof of concept and created a [kwctl release](https://github.com/raulcabello/kwctl/releases/tag/v0.3-test-credstore) that works on my mac (and I also tested on linux with an aws credentials store). It still needs more testing and some code improvement, and unfortunately I won't be able to work on this next week. Moving to TO DO in case someone wants to pick it up while I'm away.
